### PR TITLE
Feature/keycloak

### PR DESCRIPTION
**Pull request by cthtrifork**

### CODE REVIEW

File: config/redpanda-cluster.yaml [hunk 1]
```diff
@@ -9,9 +9,10 @@ kafka:
     # username: redpanda
```
Comment (cthtrifork) [inline config/redpanda-cluster.yaml:9]: why rename

File: docker-compose/kafka.yaml [hunk 2]
```diff
@@ -89,7 +89,7 @@ services:
     environment:
       CONFIG_FILEPATH: /tmp/config.yml
     volumes:
-      - ./config/redpanda.yaml:/tmp/config.yml
+      - ./config/redpanda-admin.yaml:/tmp/config.yml
```
Comment (cthtrifork) [inline docker-compose/kafka.yaml:92]: shouldn't we use the security model as default. We could add another redpanda service with a unique profile that is called: redpanda-admin

File: config/redpanda-cluster.yaml (hunk 1 above)
Comment (pegtrifork) [reply config/redpanda-cluster.yaml:9]: To make it more clear why this config exists.

File: docker-compose/kafka.yaml (hunk 2 above)
Comment (pegtrifork) [reply docker-compose/kafka.yaml:92]: For local development I think you should have admin rights in redpanda as default. The only reason why we include the cluster config (IMO) is to be able to understand/debug why you are unable to do certain things on a more locked down cluster

File: config/redpanda-cluster.yaml (hunk 1 above)
Comment (cthtrifork) [reply config/redpanda-cluster.yaml:9]:
```suggestion
    # this file is used to simulate the security model on a cluster which uses OAuth for security
    # username: redpanda
```

File: docker-compose/kafka.yaml (hunk 2 above)
Comment (cthtrifork) [reply docker-compose/kafka.yaml:92]: I agree to a point, but opensearch and kafka is behind OAuth2 - so it is a bit inconsistent/confusing in my mind. However, the developers do not actively connect and interact with redpanda, like they do with the core services. We can let it be as-is i guess.

File: docker-compose/keycloak.yaml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    command:
+      - "start-dev"
+    environment:
+      KEYCLOAK_ADMIN: admin
+      KEYCLOAK_ADMIN_PASSWORD: admin
+      KC_HEALTH_ENABLED: true
+      KC_HOSTNAME_STRICT_BACKCHANNEL: false
+      KC_HOSTNAME: localhost
+    ports:
+      - 8080:8080
```
Comment (cthtrifork) [inline docker-compose/keycloak.yaml:13]: isn't that port almost always in use?

```suggestion
      - 1852:8080
```

**Comment by cthtrifork** [comment]
I think schema-registry client should have the following kafka roles:

```
_kafkasql-journal_read
_kafkasql-journal_write
```

**Comment by cthtrifork** [comment]
Backstage should have the single kafka role

`*_describe`

**Comment by pegtrifork** [comment]
> I think schema-registry client should have the following kafka roles:
> 
> ```
> _kafkasql-journal_read
> _kafkasql-journal_write
> ```

I agree. These have been added

**Comment by pegtrifork** [comment]
> Backstage should have the single kafka role
> 
> `*_describe`

**Comment by pegtrifork** [comment]
> Backstage should have the single kafka role
> 
> `*_describe`

Backstage is not something most delvelopers will ever spin up, which is why I didn't add this client, since this will most likely just cause confusion to most developers.
When working locally with backstage, I also think it is okay for backstage to have more rights than just describe.

File: readme.md [hunk 4]
```diff
@@ -133,10 +134,97 @@ Here is further explanation on what each profile starts.
...
+     * default_scopes: [ ] 
+     * optional_scopes:
+          - `kafka`
+               * Roles:
+                    - `Kafka_*_read`
+          - `opensearch`
+               * Roles:
+                    - `opensearch_default_read`
+- Users
+     * Description: User login via browser (See [Users](#users) for user details)
```
Comment (cthtrifork) [inline readme.md:205]: is this used ?

Comment (pegtrifork) [reply readme.md:205]: Yes. It can be used to login to login to the opensearch dashboard.

Comment (cthtrifork) [reply readme.md:205]: Should we make it a bit more explcit?

File: docker-compose/kafka.yaml [hunk 5]
```diff
@@ -164,8 +165,8 @@ services:
       REGISTRY_AUTH_ROLES_DEVELOPER: "sr-producer" # "sr-developer"
       # https://quarkus.io/guides/security-oidc-configuration-properties-reference
       quarkus.oidc.enabled: true
-      quarkus.oidc.auth-server-url: http://cheetahoauthsimulator:80
-      quarkus.oidc.jwks-path: /.well-known/jwks.json
+      quarkus.oidc.auth-server-url: http://localhost:8080/realms/local-development
```
Comment (cthtrifork) [inline docker-compose/kafka.yaml:168]: this seems invalid? don't we have a e2e which also should catch it?

Comment (cthtrifork) [reply docker-compose/kafka.yaml:168]:
```suggestion
      quarkus.oidc.auth-server-url: http://keycloak:8080/realms/local-development
```

File: readme.md [hunk 6]
```diff
@@ -133,10 +134,97 @@ Here is further explanation on what each profile starts.
...
 |       Grafana        |   x   |       |       |            |       x       |
+
+## Keycloak
+
+Keycloak is used as a local identity provider, to be able to mimic a production security model with service to service authentication.
+
+### Useful urls:
+
+- OpenID Endpoint Configuration: http://localhost:8080/realms/local-development/.well-known/openid-configuration
+- Token Endpoint http://localhost:8080/realms/local-development/protocol/openid-connect/token
```
Comment (cthtrifork) [inline readme.md:153]: wrong port

```suggestion
- Token Endpoint http://localhost:1852/realms/local-development/protocol/openid-connect/token 
```

File: docker-compose/keycloak.yaml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+services:
+  keycloak:
+    image: quay.io/keycloak/keycloak:22.0.5
```
Comment (cthtrifork) [inline docker-compose/keycloak.yaml:3]:
```suggestion
    image: quay.io/keycloak/keycloak:22.0.5
    hostname: keycloak
    container_name: keycloak
```

File: docker-compose/keycloak.yaml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    command:
+      - "start-dev"
+    environment:
+      KEYCLOAK_ADMIN: admin
+      KEYCLOAK_ADMIN_PASSWORD: admin
+      KC_HEALTH_ENABLED: true
+      KC_HOSTNAME_STRICT_BACKCHANNEL: false
+      KC_HOSTNAME: localhost
```
Comment (cthtrifork) [inline docker-compose/keycloak.yaml:11]:
```suggestion
      KC_HOSTNAME: keycloak
```

Not sure about this part. Should be tested

File: docker-compose/keycloak.yaml [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    image: quay.io/adorsys/keycloak-config-cli:5.9.0-22.0.4
+    user: root
+    entrypoint: ["/bin/bash", "-c", "mkdir -p /config && /tmp/keycloak_custom_role_creation.sh /tmp/local-development.json /config/local-development.json && exec java $JAVA_OPTS -jar /app/keycloak-config-cli.jar $0 $@"]
+    depends_on:
+      keycloak:
+        condition: service_healthy
+    volumes:
+      - ./config/keycloak:/tmp
+    environment:
+      - KEYCLOAK_URL=http://keycloak:8080
```
Comment (cthtrifork) [inline docker-compose/keycloak.yaml:30]: sometimes we are using localhost and sometimes keycloak. We are sure its not a problem?

File: readme.md [hunk 10]
```diff
@@ -133,10 +134,97 @@ Here is further explanation on what each profile starts.
...
 |      Prometheus      |   x   |       |       |            |       x       |
 |       Grafana        |   x   |       |       |            |       x       |
+
+## Keycloak
+
+Keycloak is used as a local identity provider, to be able to mimic a production security model with service to service authentication.
+
+### Useful urls:
+
+- OpenID Endpoint Configuration: http://localhost:8080/realms/local-development/.well-known/openid-configuration
```
Comment (cthtrifork) [inline readme.md:152]:
```suggestion
- OpenID Endpoint Configuration: http://localhost:1852/realms/local-development/.well-known/openid-configuration
```

File: readme.md [hunk 11]
```diff
@@ -133,10 +134,97 @@ Here is further explanation on what each profile starts.
...
+- Token Endpoint http://localhost:8080/realms/local-development/protocol/openid-connect/token 
+
+### Default clients:
+
+A set of default clients have been defined which covers most common usecases.
+
+All roles are mapped to the `roles` claim in the JWT. This configuration is defined in [local-development.json](./config/keycloak/local-development.json) and is applied to keycloak using the `keycloak-setup` service.
+To modify the configuration either go to the [admin console](http://localhost:8080/admin) (Username: `admin` Password: `admin`) or edit the `local-development.json` following this [guide](./config/keycloak/setup.md)
```
Comment (cthtrifork) [inline readme.md:160]:
```suggestion
To modify the configuration either go to the [admin console](http://localhost:1852/admin) (Username: `admin` Password: `admin`) or edit the `local-development.json` following this [guide](./config/keycloak/setup.md)
```

**Review by cthtrifork — COMMENTED** [review]
Added a few small fixes. Otherwise it looks good

File: readme.md (hunk 4 above)
Comment (cthtrifork) [reply readme.md:205]:
```suggestion
     * Description: User login via browser such as OpenSearch Dashboard (See [Users](#users) for user details)
```

Comment (pegtrifork) [reply readme.md:205]: Good idea

File: docker-compose/keycloak.yaml (hunk 9 above)
Comment (pegtrifork) [reply docker-compose/keycloak.yaml]: There were some issues with mixed ports on localhost/in docker. So I changed the internal port to 1852 as well.
But it seems to work as intended with the mix of localhost and keycloak, both when running all of the infrastructure, and a flink-job inside and outside docker.

File: docker-compose/keycloak.yaml (hunk 8 above)
Comment (pegtrifork) [reply docker-compose/keycloak.yaml:15]: Not needed

**Review by rkthtrifork — APPROVED** [review]
👍

**Pull request merged**